### PR TITLE
Complexity limiter prototype

### PR DESCRIPTION
**Pull request by rogyar**
### Description
This PR adds a query complexity limiter prototype. It consists of simple operations now:
- Limiting the number of fields in a query up to 150
- Limiting the query depth to 50

The values were set experimentally by building long queries. 

The limiter is introduced as a new class. The idea is the following, we cannot say exactly now what limits we will use in the final version. It can be determined over the time by building different queries with different complexity, adding new fields to queries etc. The very possible scenario - we will have limits for some specific queries (catalog, orders). So a separate limiting logic will be developed for different scenarios. That's why we need to have it in a separate place like the new class introduced in the current PR (maybe we will need separate limiter services in the future, who knows).

Current limiting values are declared in the `di.xml` file in order to change the values easily in the future.

### Fixed Issues (if relevant)
1. https://github.com/magento/graphql-ce/issues/28 : Implement query complexity limiting

### Manual testing scenarios
1. Create a separate category and some products assigned to the category
2. Use the following crazy queries to test complexity and depth limiting (replace category_id with the actual value):

**Complexity**

```
{
category(id: 2) {
    products {
      items {
        name
        categories {
          id
          position
          level
          url_key
          url_path
          product_count
          breadcrumbs {
            category_id
            category_name
            category_url_key
          }
          products {
            items {
              media_gallery_entries {
                file
              }
              name
              special_from_date
              special_to_date
              new_to_date
              new_from_date
              tier_price
              manufacturer
              thumbnail
              sku
              image
              canonical_url
              updated_at
              created_at
              categories {
                id
                position
                level
                url_key
                url_path
                product_count
                breadcrumbs {
                  category_id
                  category_name
                  category_url_key
                }
                products {
                  items {
                    name
                    special_from_date
                    special_to_date
                    new_to_date
                    thumbnail
                    new_from_date
                    tier_price
                    manufacturer
                    sku
                    image
                    canonical_url
                    updated_at
                    created_at
                    media_gallery_entries {
                      position
                      id
                      types
                    }
                    categories {
                      id
                      position
                      level
                      url_key
                      url_path
                      product_count
                      breadcrumbs {
                        category_id
                        category_name
                        category_url_key
                      }
                      products {
                        items {
                          name
                          special_from_date
                          special_to_date
                          new_to_date
                          new_from_date
                          tier_price
                          manufacturer
                          thumbnail
                          sku
                          image
                          canonical_url
                          updated_at
                          created_at
                          categories {
                            id
                            position
                            level
                            url_key
                            url_path
                            product_count
                            breadcrumbs {
                              category_id
                              category_name
                              category_url_key
                            }
                            products {
                              items {
                                name
                                special_from_date
                                special_to_date
                                new_to_date
                                new_from_date
                                tier_price
                                manufacturer
                                sku
                                image
                                canonical_url
                                updated_at
                                created_at
                                categories {
                                  id
                                  position
                                  level
                                  url_key
                                  url_path
                                  product_count
                                  breadcrumbs {
                                    category_id
                                    category_name
                                    category_url_key
                                  }
                                  products {
                                    items {
                                      name
                                      special_from_date
                                      special_to_date
                                      new_to_date
                                      new_from_date
                                      tier_price
                                      manufacturer
                                      sku
                                      image
                                      thumbnail
                                      canonical_url
                                      updated_at
                                      created_at
                                      categories {
                                        id
                                        position
                                        level
                                        url_key
                                        url_path
                                        product_count
                                        default_sort_by
                                        breadcrumbs {
                                          category_id
                                          category_name
                                          category_url_key
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

**Depth**

```
{
  category(id: 2) {
    products {
      items {
        name
        categories {
          products {
            items {
              media_gallery_entries {
                file
              }
              categories {
                products {
                  items {
                    categories {
                      products {
                        items {
                          categories {
                            products {
                              items {
                                categories {
                                  products {
                                    items {
                                      categories {
                                        products {
                                          items {
                                            categories {
                                              products {
                                                items {
                                                  categories {
                                                    products {
                                                      items {
                                                        categories {
                                                          products {
                                                            items {
                                                              name,
                                                              categories {
                                                                products {
                                                                  items {
                                                                    categories {
                                                                      products {
                                                                        items {
                                                                          categories{
                                                                            products {
                                                                              items {
                                                                                categories {
                                                                                  products {
                                                                                    items {
                                                                                      categories {
                                                                                        products {
                                                                                          items {
                                                                                            categories {
                                                                                              products {
                                                                                                items {
                                                                                                  name,
                                                                                                  categories {
                                                                                                    products {
                                                                                                      items {
                                                                                                        categories {
                                                                                                          name
                                                                                                        }
                                                                                                      }
                                                                                                    }
                                                                                                  }
                                                                                                }
                                                                                              }
                                                                                            }
                                                                                          }
                                                                                        }
                                                                                      }
                                                                                    }
                                                                                  }
                                                                                }
                                                                              }
                                                                            }
                                                                          }
                                                                        }
                                                                      }
                                                                    }
                                                                  }
                                                                }
                                                              }
                                                            }
                                                          }
                                                        }
                                                      }
                                                    }
                                                  }
                                                }
                                              }
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

### CODE REVIEW

**Comment by rogyar** [comment]
@paliarush Hi Alex. I've reduced the values to 
Complexity: 50
Depth: 10

I've tested with different queries and it sounds reasonable to me. What do you think? 
Thank you.

**Comment by paliarush** [comment]
@rogyar Introspection does not work for me in case when depth limiting is set to 10. It was fine in the past because depth limiting was disabled in developer mode, but now it can be a problem. Please check that use case.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Framework/QueryComplexityLimiterTest.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            }
+          }
+        }
+      }
+    }
+  }
+}
+QUERY;
+
+        self::expectExceptionMessageRegExp('/Max query complexity should be 50 but got 62/');
```
Comment (paliarush) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Framework/QueryComplexityLimiterTest.php]: Would be nice to reference documentation about query complexity from here. E.g. to this one https://github.com/webonyx/graphql-php/blob/master/docs/security.md

File: lib/internal/Magento/Framework/GraphQl/Query/QueryProcessor.php [hunk 2]
```diff
@@ -24,12 +21,21 @@ class QueryProcessor
      */
     private $exceptionFormatter;
 
+    /**
+     * @var QueryComplexityLimiter
+     */
+    protected $queryComplexityLimiter;
```
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Query/QueryProcessor.php]: Please use `private` visibility by default.

**Review by paliarush — APPROVED** [review]

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Framework/QueryComplexityLimiterTest.php (hunk 1 above)
Comment (rogyar) [reply dev/tests/api-functional/testsuite/Magento/GraphQl/Framework/QueryComplexityLimiterTest.php]: I've added the reference to the `\Magento\Framework\GraphQl\Query\QueryComplexityLimiter` class DocBlock

**Comment by rogyar** [comment]
I've just fixed the failing tests. Let's see what travis says :)

File: lib/internal/Magento/Framework/GraphQl/Query/QueryProcessor.php [hunk 3]
```diff
@@ -49,10 +55,11 @@ public function process(
         array $variableValues = null,
         string $operationName = null
     ) : array {
+        var_dump($this->exceptionFormatter->shouldShowDetail());
```
Comment (naydav) [inline lib/internal/Magento/Framework/GraphQl/Query/QueryProcessor.php]: Function or method 'var_dump()' is obsolete in file '/opt/bamboo/agents/agent1/xml-data/build-dir/MCCE23-L41453-SL/magento2/lib/internal/Magento/Framework/GraphQl/Query/QueryProcessor.php'.

**Comment by naydav** [comment]
Pls, check test
```

BundleProductViewTest testAndMaxMinPriceBundleProduct History | REST EE | < 1 sec
-- | -- | --
Magento\GraphQl\Bundle\BundleProductViewTest::testAndMaxMinPriceBundleProduct Exception: JSON Encoding / Decoding error: Syntax error, invalid JSON'bool(false)
```

**Review by naydav — CHANGES_REQUESTED** [review]

**Review by paliarush — APPROVED** [review]

**Comment by naydav** [comment]
We gave got failed PAT build
https://bamboo.corp.magento.com/browse/MCCE23-PAT1279-5
Need to more time for investigate

**Review by paliarush — APPROVED** [review]

**Review by paliarush — APPROVED** [review]

**Pull request merged**